Define constrained-FDT trailer and whole-FIT signing (#51)
open-source-firmware/flat-image-tree: source/chapter7-security.rst

is hashed with SHA-256. The resulting digest is then signed with the RSA-2048
private key to produce the signature ``value``.

Whole-FIT signing
-----------------

Per-configuration signing covers the integrity of one configuration and

> the images it references. A FIT may additionally carry a *whole-FIT

Reviewer: ```suggestion
the images it references. A FIT can also carry a *whole-FIT
```

To avoid the misunderstanding that Whole-FIT requires separately signed configurations as well

> signature* over the entire FIT image, including the FDT and any

Reviewer 2: ```suggestion
signature* over the entire FIT, including the FDT and any
```

external image data. Whole-FIT signing is intended to complement
per-configuration signing, not to replace it.

A whole-FIT signature is stored in the binary trailer (see
:ref:`chapter-binary-trailer`) as a ``signature-N`` property. The
trailer is a constrained FDT blob placed inside the main FDT but
excluded from the signed range, so its contents (including the
signatures themselves) can be modified post-signing.

Why whole-FIT signing
~~~~~~~~~~~~~~~~~~~~~

Per-configuration signing requires the bootloader to parse the FIT's
FDT before authentication: it must locate the configuration node, the
referenced image nodes, and the signature node, then construct the
node list and hash the relevant FDT structure bytes. The pre-auth
parser is therefore a full FDT parser (e.g. libfdt), which is large
and has historically been a source of bugs when fed adversarial
input.

Whole-FIT signing allows a bootloader to authenticate the FIT before

> parsing any FDT structure. The verifier reads a small number of

Reviewer: This is not correct any longer, because the trailer is FDT-formatted. Yes, it's a simpler subset, but this sentence is not correct.

fixed-offset fields, hashes a contiguous byte range, and checks the
signature. Only after the signature verifies does the bootloader
parse the FDT — and at that point it is operating on trusted bytes.

Bootloaders that need a smaller pre-authentication attack surface
should use whole-FIT signing. Bootloaders that already trust the FDT

> parser they use may continue to use per-configuration signing alone.

Reviewer: I also think that the Whole-FIT approach is a better idea logic-bug wise. Determining what exactly is covered by the signature/hash is much more complex for configuration signing that for whole FIT signing and it has been a source of bugs in the past.

A FIT may carry both, so a single image can serve both kinds of
bootloader.

What is signed
~~~~~~~~~~~~~~

A whole-FIT signature covers two byte ranges, in this order:

#. ``[0, 0x28)`` — the main FDT header (40 bytes). The trailer's fixed
   ``totalsize`` keeps every header field stable, so the entire header
   is signed without exclusions. This authenticates the FDT's
   internal layout (``totalsize``, ``off_dt_struct``, ``off_dt_strings``,
   ``off_mem_rsvmap``).

> #. ``[0x28 + trailer_totalsize, totalsize + external_data_size)`` —

Reviewer 2: While I like the idea of authenticating the FIT as a whole, unconditionally covering also all external data seems redundant and not even a good idea in every case -- after all, the FIT contains hashes (or a dt-verity root hash and parameters) for all sub-images. Verifying all external data can take quite a long time, and in U-Boot context would typically require loading everything into RAM.
One could argue that you can still set `external_data_size` to 0 while actual external data is present, and then (if needed elsewhere) calculate the actual on-disk size of FIT + (only implicitly signed) external data as done [here](https://github.com/openwrt/openwrt/blob/b94de14bafd06660536691ed633f364edf5fbe4d/package/boot/uboot-mediatek/patches/200-cmd-add-imsz-and-imszb.patch#L86).
However, imho it would be good to make it clear that `external_data_size` in the trailer means size of *signed* external data, and there may also be *implicitly signed* external data not covered by the "Whole FIT Signature". Ie. maybe name the trailer attribute `signed_external_data_size`?

Reviewer: Whole-FIT signing is about signing the FIT as a whole and thus avoiding much of the complexity of the FIT format. I would like to avoid adding yet another signing scheme that only covers some parts of the image only to bite us later.

My idea is that external_data_size can be used in calculating the total FIT size independently of whether Whole-FIT is used or not.
